ur_robot_driver/src/ros/dashboard_client.cpp

```diff
@@ -0,0 +1,130 @@
+// this is for emacs file handling -*- mode: c++; indent-tabs-mode: nil -*-
+
+// -- BEGIN LICENSE BLOCK ----------------------------------------------
+// Copyright 2019 FZI Forschungszentrum Informatik
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// -- END LICENSE BLOCK ------------------------------------------------
+
+//----------------------------------------------------------------------
+/*!\file
+ *
+ * \author  Felix Mauch mauch@fzi.de
+ * \date    2019-10-21
+ *
+ */
+//----------------------------------------------------------------------
+
+#include <ur_robot_driver/ros/dashboard_client.h>
+
+namespace ur_driver
+{
+DashboardClientROS::DashboardClientROS(const ros::NodeHandle& nh, const std::string& robot_ip)
+  : nh_(nh), client_(robot_ip, DASHBOARD_SERVER_IP)
+{
+  client_.connect();
+  brake_release_service_ = nh_.advertiseService("brake_release", &DashboardClientROS::brakeRelease, this);
+  clear_operational_mode_service_ =
+      nh_.advertiseService("clear_operational__mode", &DashboardClientROS::clearOperationalMode, this);
+  close_popup_service_ = nh_.advertiseService("close_popup", &DashboardClientROS::closePopup, this);
+  close_safety_popup_service_ = nh_.advertiseService("close_safety_popup", &DashboardClientROS::closeSafetyPopup, this);
+  pause_service_ = nh_.advertiseService("pause", &DashboardClientROS::pause, this);
+  play_service_ = nh_.advertiseService("play", &DashboardClientROS::play, this);
+  power_off_service_ = nh_.advertiseService("power_off", &DashboardClientROS::powerOff, this);
+  power_on_service_ = nh_.advertiseService("power_on", &DashboardClientROS::powerOn, this);
+  quit_service_ = nh_.advertiseService("quit", &DashboardClientROS::quit, this);
+  restart_safety_service_ = nh_.advertiseService("restart_safety", &DashboardClientROS::restartSafety, this);
+  shutdown_service_ = nh_.advertiseService("shutdown", &DashboardClientROS::shutdown, this);
+  stop_service_ = nh_.advertiseService("stop", &DashboardClientROS::stop, this);
+  unlock_protective_stop_service_ =
+      nh_.advertiseService("unlock_protective_stop", &DashboardClientROS::unlockProtectiveStop, this);
```

Comment:
Could consider using lambdas here. Most service callbacks are single line 'forwards' to the `client_` member.

IIRC, lambdas should be usable as service server callbacks.

Reply:
I changed this in the last commit to make the actual client very slim as interpreting the server's answer at such a low level seemed wrong to me. This makes this kind of obsolete, I guess.

Reply:
I'm not sure. There still appear to be many very simple service callbacks that could be changed to lambdas instead.

Reply:
You mean like that?

```c++
  // Service to release the brakes. If the robot is currently powered off, it will get powered on on the fly.
  brake_release_service_ = nh_.advertiseService<std_srvs::Trigger::Request, std_srvs::Trigger::Response>(
      "brake_release", [&](std_srvs::Trigger::Request& req, std_srvs::Trigger::Response& resp) {
        resp.message = client_.sendAndReceive("brake release\n");
        resp.success = std::regex_match(resp.message, std::regex("Brake releasing"));
        return true;
      });
```

Or do I miss a way of simplifying this even more?

Reply:
Yes, for instance. And seeing as it's basically the same pattern over and over, you may introduce a macro which takes a few arguments. That could reduce all of this to single line definitions for every service server.